pynautobot/core/endpoint.py

```diff
@@ -311,42 +310,101 @@ def create(self, *args, api_version=None, **kwargs):
-    def update(self, id: str, data: Dict[str, any]):
-        """
-        Update a resource with a dictionary.
+    def update(self, *args, **kwargs):
+        r"""
+        Update a single resource with a dictionary or bulk update a list of objects.
+
+        Allows for bulk updating of existing objects on an endpoint.
+        Objects is a list which contain either json/dicts or Record
+        derived objects, which contain the updates to apply.
+        If json/dicts are used, then the id of the object *must* be
+        included
 
-        Accepts the id of the object that needs to be updated as well as
-        a dictionary of k/v pairs used to update an object. The object
-        is directly updated on the server using a PATCH request without
-        fetching object information.
+        :arg str,optional \*args: A list of dicts or a list of Record
 
-        For fields requiring an object reference (such as a device location),
-        the API user is responsible for providing the object ID or the object
-        URL. This API will not accept the pynautobot object directly.
+        :arg str,optional \**kwargs:
+            See Below
 
-        :arg str id: Identifier of the object being updated
-        :arg dict data: Dictionary containing the k/v to update the
-            record object with.
-        :returns: True if PATCH request was successful.
+        :Keyword Arguments:
+            * *id* (``string``) -- Identifier of the object being updated
+            * *data* (``dict``) -- k/v to update the record object with
+
+        :returns: A list or single :py:class:`.Record` object depending
+            on whether a bulk update was requested.
         :example:
 
+        Accepts the id of the object that needs to be updated as well as a
+            dictionary of k/v pairs used to update an object
         >>> nb.dcim.devices.update(id="0238a4e3-66f2-455a-831f-5f177215de0f", data={
-        ...     "name": "test-switch2",
-        ...     "serial": "ABC321",
+        ...     "name": "test",
+        ...     "serial": "1234",
         ...     "location": "9b1f53c7-89fa-4fb2-a89a-b97364fef50c",
         ... })
-        True
+        >>>
+
+        Use bulk update by passing a list of dicts:
+
+        >>> devices = nb.dcim.devices.update([
+        ...    {'id': "db8770c4-61e5-4999-8372-e7fa576a4f65", 'name': 'test'},
+        ...    {'id': "e9b5f2e0-4f20-41ad-9179-90a4987f743e", 'name': 'test2'},
+        ... ])
+        >>>
+
+        Use bulk update by passing a list of Records:
+
+        >>> devices = list(nb.dcim.devices.filter())
+        >>> devices
+        [Device1, Device2, Device3]
+        >>> for d in devices:
+        ...     d.name = d.name+'-test'
+        ...
+        >>> nb.dcim.devices.update(devices)
+        >>>
         """
+        objects = args[0] if args else []
+        if not isinstance(objects, list):
+            raise ValueError("objects must be a list[dict()|Record] not " + str(type(objects)))
```

Comment:
I don't think this would be fully backwards compatible. In the current version, you could technically send the arguments in as positional args instead of keyword args and it would break in this version.

For example, this is technically possible today:
```
nb.dcim.devices.update("0238a4e3-66f2-455a-831f-5f177215de0f", data={"foo": "bar"})
```

Please account for this by checking `len(args)` and splitting them accordingly if the length is 2.

Reply:
I prefer to keep the status quo with backwards compatibility. If you can incorporate that by using the same method then great, but I'm fine with creating a "bulk" method as well. It does draw the question if we should then rename `.delete()` for consistency (as it hasn't been released yet).

Reply:
Keeping it incorporated with the same method, I came up with this, and it passes those cases.  At some point, the logic inside the method could become too complicated for backwards compatibility.  What do you think about this approach?

```
uuid = ""
data = {}
objects = []
if len(args) == 1 and not kwargs:
    objects = args[0]
    data = []
elif len(args) == 1 and "data" in kwargs:
    uuid = args[0]
    data = kwargs["data"]
elif len(args) >= 2:
    uuid = args[0]
    data = args[1]
elif "data" in kwargs and "id" in kwargs:
    uuid = kwargs["id"]
    data = kwargs["data"]
    
---omitted---
```

Reply:
I think this might look cleaner. Please verify this catches all of the use cases.
```python
    def update(self, *args, **kwargs):
        # Original signature: update(self, id: str, data: Dict[str, any])
        if not args and not kwargs:
            # If we got no args or kwargs, raise a ValueError
            raise ValueError("You must provide either a UUID and data dict or a list of objects to update")
        uuid = kwargs.get("id", "")
        data = kwargs.get("data", {})
        if data and not uuid:
            # If we got data but no uuid, the uuid should be the first arg
            uuid = args[0]
        if len(args) == 2:
            # If we got two args, the first should be the uuid and the second should be the data
            uuid, data = args
        if uuid and data:
            # If we have both, dispatch the original single update method
            < ... >

        # If we got here, we should be dealing with a bulk update
        < ... >
```

Reply:
Combining what you have with some changes I made to reduce code redundancy, it's looking like this.  All test cases are passing - 

```
        if not args and not kwargs:
            raise ValueError("You must provide either a UUID and data dict or a list of objects to update")
        uuid = kwargs.get("id", "")
        data = kwargs.get("data", {})
        if data and not uuid:
            uuid = args[0]
        if len(args) == 2:
            uuid, data = args
        
        if not uuid and not data:
            objects = args[0] if args else []
            bulk_data = []

            if not isinstance(objects, list):
                raise ValueError("objects must be a list[dict()|Record] not " + str(type(objects)))

            for o in objects:
                try:
                    if isinstance(o, dict):
                        bulk_data.append(o)
                    elif isinstance(o, Record):
                        if not hasattr(o, "id"):
                            raise ValueError("'Record' object has no attribute 'id'")
                        updates = o.updates()
                        if updates:
                            updates["id"] = o.id
                            bulk_data.append(updates)
                    else:
                        raise ValueError("Invalid object type: " + str(type(o)))
                except ValueError as exc:
                    raise ValueError("Unexpected value in object list") from exc

        req = Request(
            key=uuid,
            base=self.url,
            token=self.api.token,
            http_session=self.api.http_session,
            api_version=self.api.api_version,
        )
        if uuid and data:
            if req.patch(data):
                return True
            return False
        if objects and bulk_data:
            return response_loader(req.patch(bulk_data), self.return_obj, self)
```

Reply:
I have a better idea, let's add the bulk update method and call it via the update method.
```python
    def update(self, *args, **kwargs):
        # Original signature: update(self, id: str, data: Dict[str, any])
        if not args and not kwargs:
            raise ValueError("You must provide either a UUID and data dict or a list of objects to update")
        uuid = kwargs.get("id", "")
        data = kwargs.get("data", {})
        if data and not uuid:
            uuid = args[0]
        if len(args) == 2:
            uuid, data = args

        if not any([uuid, data]):
            return self.bulk_update(args[0])

        req = Request(
            key=uuid,
            base=self.url,
            token=self.api.token,
            http_session=self.api.http_session,
            api_version=self.api.api_version,
        )
        if req.patch(data):
            return True
        return False

    def bulk_update(self, objects: List[Dict[str, Any]]):
        if not isinstance(objects, list):
            raise ValueError("objects must be a list[dict()|Record] not " + str(type(objects)))

        bulk_data = []
        for o in objects:
            try:
                if isinstance(o, dict):
                    bulk_data.append(o)
                elif isinstance(o, Record):
                    if not hasattr(o, "id"):
                        raise ValueError("'Record' object has no attribute 'id'")
                    updates = o.updates()
                    if updates:
                        updates["id"] = o.id
                        bulk_data.append(updates)
                else:
                    raise ValueError("Invalid object type: " + str(type(o)))
            except ValueError as exc:
                raise ValueError("Unexpected value in object list") from exc

        req = Request(
            base=self.url,
            token=self.api.token,
            http_session=self.api.http_session,
            api_version=self.api.api_version,
        ).patch(bulk_data)
        return response_loader(req, self.return_obj, self)
```

Reply:
I like this approach much better!